db/versions/0091-migration.sql

```diff
@@ -0,0 +1,183 @@
+    pop_receipt uuid null
+  );
+
+  -- migrate data to claimed queue
+  INSERT INTO
+    queue_claimed_tasks (task_id, run_id, task_queue_id, worker_group, worker_id, claimed, taken_until, visible, pop_receipt)
+  SELECT
+    convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'taskId',
+    CAST(convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'runId' AS INTEGER),
+    '', -- task_queue_id was not present in old format
+    '', -- worker_group was not present in old format
+    '', -- worker_id was not present in old format
+    inserted,
+    CAST(convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'takenUntil' AS timestamp with time zone),
+    visible,
+    pop_receipt
+  FROM azure_queue_messages
+  WHERE queue_name = 'claim-queue'
+  AND expires > now();
+
+  -- before we could add unique (task_id, run_id) we need to ensure we only keep the latest record in the table
+  -- delete all but the latest taken_until
+  DELETE FROM queue_claimed_tasks
+  WHERE (task_id, run_id, taken_until) NOT IN (
+    SELECT task_id, run_id, MAX(taken_until)
+    FROM queue_claimed_tasks
+    GROUP BY task_id, run_id
+  );
+
+  -- we could only have single claim-expire for given run_id
+  -- some workers might reclaim tasks more frequently,
+  -- so there could be multiple records fro the same run
+  CREATE UNIQUE INDEX queue_claimed_task_run_idx ON queue_claimed_tasks (task_id, run_id);
+  CREATE INDEX queue_claimed_task_vis_idx ON queue_claimed_tasks (visible);
+  CREATE INDEX queue_claimed_task_queue_idx ON queue_claimed_tasks (task_queue_id, worker_group, worker_id);
+
+
+  -- Pending (scheduled)
+  -- A task appears in this table if it is pending,
+  -- but a non-pending task may also appear in this table.
+  CREATE TABLE queue_pending_tasks (
+    task_queue_id text not null,
+    priority int not null,
+    task_id text not null, -- should be unique and a key
+    run_id integer not null,
+    hint_id text not null, -- to know that task was claimed properly, stored inside tasks.runs[]
+    inserted timestamptz not null,
+    expires timestamptz not null,
+    visible timestamptz not null,
+    pop_receipt uuid null
+  );
+
+  -- rest goes into queue_pending_tasks
+  INSERT INTO queue_pending_tasks
+    (task_queue_id, priority, task_id, run_id, hint_id, inserted, expires, visible, pop_receipt)
+  SELECT
+    task_queue_id,
+    priority,
+    convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'taskId',
+    CAST(convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'runId' AS INTEGER),
+    convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'hintId',
+    inserted,
+    expires,
+    visible,
+    pop_receipt
+  FROM azure_queue_messages
+  WHERE queue_name NOT IN ('claim-queue', 'deadline-queue', 'resolved-queue')
+  AND task_queue_id IS NOT NULL
+  AND expires > now();
+
+  CREATE UNIQUE INDEX queue_pending_task_idx ON queue_pending_tasks (task_id);
```

Comment:
@djmitche, @petemoore as I was reading older comments and docs I noticed that before it was allowed to have multiple  `(taskId, runId)` messages in the queue. 
With this change I'm enforcing "there can be only one run scheduled for any task".

Do you think there are any reasons or conditions under which several runs for the same task could be pending? 
I believe this was already enforced in `schedule_task` and `rerun_task` methods.

So I think it is fine that we always keep single record here for any task to let workers claim only the latest run

Reply:
Historically, queues have been "over-estimations", so for example a task being in the pending queue meant that the task _might_ actually be pending. But a task that was not in the pending queue was definitely not pending. This is a pretty common distributed-algorithm design, but was necessary with azure queues since it's impossible to delete a message from a queue - or know if a message exists in a queue - until that message is visible and comes to the top of the queue.

So I think a (task_id, run_id) could appear in the queue multiple times if there are some overlaps or failed operations, and might prevent retrying an idempotent operation. I can't think of a case in "normal operation" where that might occur. So preventing this would change the semantics of the queue and _might_ have undesirable consequences.

I think it's a judgement call on your part whether to change the semantics for some, or all, of the queues. And if you change it, whether to do so in this PR or a followup. The cost to changing this is complexity of the PRs and the possibility of bugs. Those bugs would be of the sort where a pending task never runs, or a worker failure does not result in `claim-expired`, or something like that -- the sort of bug that is rare and hard to track down.

If you choose to keep the existing queue semantics, this PR is already a big improvement: values are broken out into columns that can be manipulated with SQL. For example, if you allow that multiple pending messages for the same run of the same task can exist in the pending queue, you can still find all pending tasks for a taskQueueId with something like
```sql
SELECT * from queue_tasks JOIN (
  SELECT unique task_id
  FROM queue_pending_tasks
  WHERE task_queue_id = $taskQueueId
) as maybe_pending
WHERE queue_tasks.task_id = maybe_pending.task_id
AND queue_tasks.status = 'pending'     -- I forgot how this status is stored, but you get the idea
```

Reply:
This is helpful. I prefer not to change the semantics now as I don't have a complete understanding about the impacts. 

Agree, I'd keep it for now as it was before to reduce the risk of having unwanted side-effects during this migration.